### PR TITLE
Add ability to set a classical condition on a list of bits

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Use list of classical bits in conditions

### Details and comments

This PR is a continuation of #6761 by @TharrmashasthaPV. That PR created all the underlying code to represent conditions as a list of bits. This PR extends this by modifying the 3 circuit drawers to properly display conditions that are lists. For example,
```
clbits = [Clbit(), Clbit(), Clbit()]
qr = QuantumRegister(2, "qr")
cr = ClassicalRegister(3, "cr")
circuit = QuantumCircuit(qr, clbits, cr)
circuit.x(0).c_if([cr[0], clbits[1], cr[1], clbits[2]], 11)
circuit.draw("mpl", cregbundle=False)
```
image
As part of this, if `cregbundle` is on and at least one of the bits in the list belongs to a register, then `cregbundle` is turned off with a warning to the user. This allows full display of the details of the condition.

Also now all conditions in the 'mpl' drawer display the hex value at the bottom to be consistent.

### CODE REVIEW

File: qiskit/visualization/matplotlib.py
Comment (javabster) [inline qiskit/visualization/matplotlib.py:414]: This snippet looks very similar to what's in `latex.py`, would it make sense to move this out to a utils file and reduce duplication of logic?

File: qiskit/visualization/utils.py
Comment (javabster) [inline qiskit/visualization/utils.py:284]: Maybe beyond the scope of this PR, but I think as this utils file keeps growing it would be good at some point to separate these helper functions out into multiple files, in their own utils folder

File: qiskit/visualization/matplotlib.py
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:414]: Actually basically matches the 'text' drawer code as well, though it handles the `cargs` version differently. Could definitely be added to utils. Will look into it.

File: qiskit/visualization/utils.py
Comment (javabster) [reply qiskit/visualization/utils.py:284]: That would be great! And it would be great to get an architecture diagram put together to go with with that, I feel like the visualisation code can be quite confusing for newcomers so something like that would help a lot

Comment (javabster) [reply qiskit/visualization/utils.py:284]: again, beyond the scope of this PR but food for thought in future 😄

File: qiskit/visualization/matplotlib.py
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py]: @javabster Moved the `cregbundle` checks to `utils.py` in https://github.com/Qiskit/qiskit-terra/pull/7653/commits/829ded3621a87d63b7e4dca8fb0aa180f65e15c1.

**Comment by enavarro51** [comment]
@javabster The `cregbundle `checks are now in `utils`. The mpl drawer goes through the nodes before drawing in order to get layer widths. I added the checks into this for loop. The latex drawer also goes through the nodes at the beginning to get `image_width`, and I added the checks there. The text drawer does not have a pre-processing loop, so I had to make an extra trip through the nodes to do the `cregbundle `check. Should be ready to go now.

**Review by javabster — DISMISSED** [review]
This LGTM, thanks Ed! I think @mtreinish might have some thoughts on the dagcircuit changes?

File: qiskit/circuit/instruction.py
Comment (1ucian0) [inline qiskit/circuit/instruction.py:430]: I tried to trigger this exception (in order to create a test for it). But it seems like it is covered by line 421 
> raise CircuitError("condition value should be non-negative")

Comment (1ucian0) [reply qiskit/circuit/instruction.py:430]: Triggered! testing in 9cafc8571

Still not fully sure how 421 does not get this one.

File: qiskit/circuit/instruction.py
Comment (1ucian0) [inline qiskit/circuit/instruction.py:427]: tested in 9a29cfc31e990cedc4f8e060165b7a7caee16418

File: qiskit/visualization/text.py
Comment (1ucian0) [inline qiskit/visualization/text.py:689]: These nested loops are a bit hard to follow. I think it is  a case of [for/else](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops)

**Review by 1ucian0 — APPROVED** [review]
Thanks @enavarro51 ! I added some error-handling tests for completeness. Let's merge it!

**Comment by jakelishman** [comment]
I've been away for quite a while, so never quite got time to comment on this before it merged.  Sorry that this is coming a bit out of nowhere now.

This PR has caused relatively serious regressions in several conversion functions, though those are at least overshadowed by other factors over a complete transpilation workload.  Having the visualisations prepared to work with conditions on several bits is certainly very useful, but the problem is that this idea of conditioning on a list of bits like this is not how we'll be moving forwards with the more complex conditionals used in dynamic circuits.  This PR only adds support for this style of condition to the visualisations, but the chance of it being added to any simulator is very low, and certainly won't be in the 0.20 release.

I think it might be best to roll back this PR, rather than release known regressions into 0.20 for a feature that doesn't have a use yet.  The concept of "conditioning on a list of bits" will exist in the future, and I think a large amount of this PR will be useful (especially the visualisation components), but I think it's best not to extend the `(bit, value)` 2-tuple when it's something that we'll want to be deprecating.  Simulator support for conditionals on lists of bits won't be available until the classical type system is in place, so I think it may be better to hold off on this until we're ready to do it a bit more completely.

**Comment by jakelishman** [comment]
This idea of "conditioning on a list of bits" is actually already supported by Terra, it's just not supported on ad-hoc lists like this PR does.  You can currently just add a register to the circuit that contains bits from other registers, and apply the condition on that.

Perhaps the best course of action is to roll back the changes to `Instruction` that attempt to add list of bits conditionals, and keep the visualisation components.  We could update the visualisation components so they work when applied to a register that is not one of the drawn ones, or one that is not defined over contiguous clbits.

The changes to `Instruction.condition` here are a bit inconsistent, as well.  You can condition on a single bit by integer or instance (`c_if(0)` or `c_if(bits[0])`), but you can only condition on a list of bit instances:
```python

**Comment by jakelishman** [comment]
To be clear (and sorry for the third long message): I think we should roll back this PR, but then use the new visualiation logic to make this (currently erroring) code work:
```python
from qiskit.circuit import QuantumCircuit, ClassicalRegister, Clbit, QuantumRegister

clbits = [Clbit() for _ in [None] * 5]
cr1 = ClassicalRegister(bits=clbits)
cr2 = ClassicalRegister(bits=clbits[::2])
qc = QuantumCircuit(QuantumRegister(5), cr1, cr2)
for i in range(5):
    qc.h(i)
qc.measure(range(5), range(5))
qc.x(0).c_if(cr2, 2**3 - 1)
qc.draw()  # ValueError in Layer.set_clbit due to failed register lookup
```
This shows how "lists of bits" conditions could already be emulated (just not in an ad-hoc manner, and still with no simulator support), and the new visualisation changes I think can be used to make this work in the drawers, which is very useful.

**Comment by chriseclectic** [comment]
@jakelishman This PR is also causing test failures on qiskit-experiments, I confirmed reverting this commit fixed the failing tests.

**Pull request merged**
